docs/SQL/extensibility/gem-builder/gem-builder.md

```diff
@@ -0,0 +1,376 @@
+           )
+       )
+
+```
+
+After defining a Gem in the code editor, you can preview and test it. See [Preview](#3-preview). This feature directly renders the interface for the selected Gem using a dummy schema, enabling you to configure and experiment with the Gem’s UI components. You can then finalize them by previewing the generated SQL code.
+
+![Gem builder preview](img/gem-builder-preview.png)
+
+There are various UI components that can be defined for custom Gems such as scroll boxes, tabs, and buttons. These UI components can be grouped together in various types of panels to create a custom user experience when using the Gem.
+
+After the Dialog object is defined, it's serialized as JSON, sent to the UI, and rendered there.
+
+Depending on what kind of Gem is being created, a `Dialog` needs to be defined.
+
+#### Column selector
+
+You can use the column selector property if you want to select the columns from UI and then highlight the used columns using the `onChange` function.
+It is recommended to try out this dialogue code in Gem builder UI and see how each of these elements looks in UI.
+
+### Validation
+
+The `validate` method performs validation checks so that in the case where there's any issue with any inputs provided for the user an Error can be displayed. You can add any validation on your properties.
+
+- Optional functions such as `onChange` or `validate`, which are executed on user actions. They can dynamically alter the state of how the Gem works based on the user input.
+
+```sql
+
+    def validate(self, context: SqlContext, component: Component) -> List[Diagnostic]:
+        # Validate the component's state
+        return super().validate(context,component)
+
+```
+
+### State Changes
+
+The `onChange` method is given for the UI State transformations. You are given both the previous and the new incoming state and can merge or modify the state as needed. The properties of the Gem are also accessible to this function, so functions like selecting columns, etc. are possible to add from here.
+
+```sql
+
+    def onChange(self, context: SqlContext, oldState: Component, newState: Component) -> Component:
+        # Handle changes in the component's state and return the new state
+        return newState
+
+```
+
+### Apply
+
+The code for the Gem logic is defined in the `apply` function. Here the above User Defined properties are accessible using `self.projectName.{self.name}`.
+
+```sql
+
+    def apply(self, props: macro_gemProperties) -> str:
+        # generate the actual macro call given the component's state
+        resolved_macro_name = f"{self.projectName}.{self.name}"
+        non_empty_param = ",".join([param for param in [props.parameter1] if param != ''])
+        return f'{{{{ {resolved_macro_name}({non_empty_param}) }}}}'
+
+```
+
+### Load and Unload Properties
+
+The `loadProperties` and `unloadProperties` methods load and convert the component's state to the default macro property representation.
```

Comment:
When prophecy parses a macro invocation, it represents a macro definition in a default state: MacroProperties which consists of
 - macro name
 - project name
 - parameters used.


For example if macro invocation is 
```dbt_utils.deduplicate(relation, partition_by, order_by)```

 prophecy parses it into an object:

```MacroProperties(project_name = "dbt_utils", macro_name = "deduplicate", parameters = List(
MacroParameter(value="relation"),
MacroParameter(value="partition_by"),
MacroParameter(value="relation")
)) ```


This object now has to be converted into the gem state defined by user. This logic is defined in loadProperties. 

Similarly the the opposite case where this enhanced UX is not available due to some reason, prophecy needs to be able to render the default macro UI. For this purpose user needs to define logic to convert the gem properties back to the default macro properties object which prophecy understands.

Reply:
Done.